### ISSUE TITLE
Device/Port Association

### ISSUE THREAD

**Issue by vindicatorr** [mvp/uhubctl#368]
Motherboard specs:
1 x USB 2.0 Header (Supports 2 USB 2.0 ports)
1 x USB 3.0 Header (Supports 2 USB 3.0 ports) 
2 x USB 2.0 Ports
4 x USB 3.0 Ports

I haven't been able to see evidence of a port having power disabled to it.
I have a mass-storage and an UART device, both with LEDs so I'd know if it's supplied with power or not.
I removed my external hubs and unneeded devices.

```
$ sudo ./uhubctl
Current status for hub 2-1 [<#:#>, USB 2.00, 8 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
  Port 5: 0100 power
  Port 6: 0100 power
  Port 7: 0100 power
  Port 8: 0100 power
Current status for hub 1-1 [<#:#>, USB 2.00, 6 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
  Port 5: 0100 power
  Port 6: 0100 power
```
```
$ sudo lsusb -t
/:  Bus 04.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/6p, 5000M
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/14p, 480M
    |__ Port 3: Dev 3, If 0, Class=Human Interface Device, Driver=usbhid, 12M //wireless keyboard/mouse
    |__ Port 3: Dev 3, If 1, Class=Human Interface Device, Driver=usbhid, 12M
    |__ Port 3: Dev 3, If 2, Class=Human Interface Device, Driver=usbhid, 12M
    |__ Port 7: Dev 5, If 0, Class=Mass Storage, Driver=usb-storage, 480M //Internal front-panel reader. Not the one with LED mentioned above.
    |__ Port 11: Dev 6, If 0, Class=Wireless, Driver=btusb, 12M //internal bluetooth/wifi combo
    |__ Port 11: Dev 6, If 1, Class=Wireless, Driver=btusb, 12M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=ehci-pci/2p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/8p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=ehci-pci/2p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/6p, 480M
```
```
$ sudo lsusb -vvvv
...
Bus 002 Device 002:...
Hub Descriptor:...
    Per-port power switching
...
Hub Port Status:
   Port 1: 0000.0100 power
   Port 2: 0000.0100 power
   Port 3: 0000.0100 power
   Port 4: 0000.0100 power
   Port 5: 0000.0100 power
   Port 6: 0000.0100 power
   Port 7: 0000.0100 power
   Port 8: 0000.0100 power
...
Bus 001 Device 002:...
Hub Descriptor:...
    Per-port power switching
...
 Hub Port Status:
   Port 1: 0000.0100 power
   Port 2: 0000.0100 power
   Port 3: 0000.0100 power
   Port 4: 0000.0100 power
   Port 5: 0000.0100 power
   Port 6: 0000.0100 power
...
Bus 004 Device 001:
Hub Descriptor:...
    No power switching (usb 1.0)...
 Hub Port Status:
   Port 1: 0000.02a0 5Gbps power Rx.Detect
   Port 2: 0000.02a0 5Gbps power Rx.Detect
   Port 3: 0000.02a0 5Gbps power Rx.Detect
   Port 4: 0000.02a0 5Gbps power Rx.Detect
   Port 5: 0000.02a0 5Gbps power Rx.Detect
   Port 6: 0000.02a0 5Gbps power Rx.Detect
...
Bus 003 Device 001:...
Hub Descriptor:...
    No power switching (usb 1.0)
...
 Hub Port Status:
   Port 1: 0000.0100 power
   Port 2: 0000.0100 power
   Port 3: 0000.0103 power enable connect
   Port 4: 0000.0100 power
   Port 5: 0000.0100 power
   Port 6: 0000.0100 power
   Port 7: 0000.0503 highspeed power enable connect
   Port 8: 0000.0100 power
   Port 9: 0000.0100 power
   Port 10: 0000.0100 power
   Port 11: 0000.0103 power enable connect
   Port 12: 0000.0100 power
   Port 13: 0000.0100 power
   Port 14: 0000.0100 power
```
Of course, I don't know where all of those ports are.
It doesn't help that the devices are not under their respective Bus (hub/{8p,6p}), but under the combination (14p), whatever that is. I can associate what devices are with ports 3, 7, 11 from the lsusb tree.
I'd just assume hub 1-1 was at the rear since they both (lsusb and specs) say "6 ports".
I'd also assume that the 14p counts the 6p first since it is 1-1 and then 8p since it is 2-1

In any case, I plug in one of my LED usb devices (mass-storage)...
```
$ sudo lsusb -t
/:  Bus 04.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/6p, 5000M
    |__ Port 4: Dev 10, If 0, Class=Mass Storage, Driver=usb-storage, 5000M
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/14p, 480M
    |__ Port 3: Dev 3, If 0, Class=Human Interface Device, Driver=usbhid, 12M
    |__ Port 3: Dev 3, If 1, Class=Human Interface Device, Driver=usbhid, 12M
    |__ Port 3: Dev 3, If 2, Class=Human Interface Device, Driver=usbhid, 12M
    |__ Port 7: Dev 5, If 0, Class=Mass Storage, Driver=usb-storage, 480M
    |__ Port 11: Dev 6, If 0, Class=Wireless, Driver=btusb, 12M
    |__ Port 11: Dev 6, If 1, Class=Wireless, Driver=btusb, 12M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=ehci-pci/2p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/8p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=ehci-pci/2p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/6p, 480M
```
```
$ sudo ./uhubctl
Current status for hub 2-1 [<#:#>, USB 2.00, 8 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
  Port 5: 0100 power
  Port 6: 0100 power
  Port 7: 0100 power
  Port 8: 0100 power
Current status for hub 1-1 [<#:#>, USB 2.00, 6 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
  Port 5: 0100 power
  Port 6: 0100 power
```
```
$ sudo lsusb -vvvv
//Same entries as before except for...
Bus 004 Device 001:...
Hub Descriptor:...
    No power switching (usb 1.0)...
 Hub Port Status:
   Port 1: 0000.02a0 5Gbps power Rx.Detect
   Port 2: 0000.02a0 5Gbps power Rx.Detect
   Port 3: 0000.02a0 5Gbps power Rx.Detect
   Port 4: 0000.0203 5Gbps power U0 enable connect
   Port 5: 0000.02a0 5Gbps power Rx.Detect
   Port 6: 0000.02a0 5Gbps power Rx.Detect
```
So to turn off that "Port 4", I'm going to go with the assumption that I'd be switching it off on hub 1-1:
```
$ sudo ./uhubctl -a off -l 1-1 -p 4
$ sudo ./uhubctl
Current status for hub 2-1 [<#:#>, USB 2.00, 8 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
  Port 5: 0100 power
  Port 6: 0100 power
  Port 7: 0100 power
  Port 8: 0100 power
Current status for hub 1-1 [<#:#>, USB 2.00, 6 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0000 off
  Port 5: 0100 power
  Port 6: 0100 power
```
Nope. LED is still lit up... Sooo 2-1p4? Nope. Sooo plug it in every open port in the rear (front not used except for internal card reader). Nope. Lit up every time.

I've also been exploring <kernel>/Documentation/driver-api/usb/power-management.rst and https://microchipsupport.force.com/s/article/USB-Port-Power-Control-in-Linux-Ubuntu, but nothing came from any of it so far.

**Comment by mvp** [comment]
Very few motherboard USB controllers support per-port power switching.
However, there is one more complication that you maybe affected by.
As described here https://github.com/mvp/uhubctl#usb-30-duality-note: 
> If you have USB 3.0 hub connected to USB3 upstream port, it will be detected as 2 independent virtual hubs: USB2 and USB3, and your USB devices will be connected to USB2 or USB3 virtual hub depending on their capabilities and connection speed. To control power for such hubs, it is necessary to turn off/on power on both USB2 and USB3 virtual hubs for power off/on changes to take effect. uhubctl will try to do this automatically (unless you disable this behavior with option -e).

Keyword here `uhubctl will try to do this automatically`. Unfortunately, Linux kernel currently does not support advertising USB BOS descriptors for root USB3 hubs (motherboards typically contain only root hubs), which makes it impossible for uhubctl to do appropriate USB2 vs USB3 device pair matching.

In other words, for motherboard root USB3 hubs you have to issue 2 uhubctl commands to actually turn power off (or on), otherwise USB port may be still powered.

There are sometimes even weirder cases, when USB2 device is advertising PPPS, and matching USB3 device doesn't - it seems like your motherboard is just that. You can try using `sudo uhubctl -f` - it will list and try operating on any USB hub regardless of PPPS support, but it is very unlikely to actually work.

This is the reason why we have a table to tried and tested devices which are actually supported - just get any USB hub from that list and connect it your motherboard to get power switching support.
